internal/models/one_time_token.go

```diff
@@ -0,0 +1,330 @@
+package models
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"github.com/pkg/errors"
+	"github.com/supabase/auth/internal/storage"
+)
+
+type OneTimeTokenType int
+
+const (
+	ConfirmationToken OneTimeTokenType = iota
+	ReauthenticationToken
+	RecoveryToken
+	EmailChangeTokenNew
+	EmailChangeTokenCurrent
+	PhoneChangeToken
+)
+
+func (t OneTimeTokenType) String() string {
+	switch t {
+	case ConfirmationToken:
+		return "confirmation_token"
+
+	case ReauthenticationToken:
+		return "reauthentication_token"
+
+	case RecoveryToken:
+		return "recovery_token"
+
+	case EmailChangeTokenNew:
+		return "email_change_token_new"
+
+	case EmailChangeTokenCurrent:
+		return "email_change_token_current"
+
+	case PhoneChangeToken:
+		return "phone_change_token"
+
+	default:
+		panic("OneTimeToken: unreachable case")
+	}
+}
+
+func ParseOneTimeTokenType(s string) (OneTimeTokenType, error) {
+	switch s {
+	case "confirmation_token":
+		return ConfirmationToken, nil
+
+	case "reauthentication_token":
+		return ReauthenticationToken, nil
+
+	case "recovery_token":
+		return RecoveryToken, nil
+
+	case "email_change_token_new":
+		return EmailChangeTokenNew, nil
+
+	case "email_change_token_current":
+		return EmailChangeTokenCurrent, nil
+
+	case "phone_change_token":
+		return PhoneChangeToken, nil
+
+	default:
+		return 0, fmt.Errorf("OneTimeTokenType: unrecognized string %q", s)
+	}
+}
+
+func (t OneTimeTokenType) Value() (driver.Value, error) {
+	return t.String(), nil
+}
+
+func (t *OneTimeTokenType) Scan(src interface{}) error {
+	s, ok := src.(string)
+	if !ok {
+		return fmt.Errorf("OneTimeTokenType: scan type is not string but is %T", src)
+	}
+
+	parsed, err := ParseOneTimeTokenType(s)
+	if err != nil {
+		return err
+	}
+
+	*t = parsed
+	return nil
+}
+
+type OneTimeTokenNotFoundError struct {
+}
+
+func (e OneTimeTokenNotFoundError) Error() string {
+	return "One-time token not found"
+}
+
+type OneTimeToken struct {
+	ID uuid.UUID `json:"id" db:"id"`
+
+	UserID    uuid.UUID        `json:"user_id" db:"user_id"`
+	TokenType OneTimeTokenType `json:"token_type" db:"token_type"`
+
+	TokenHash string `json:"token_hash" db:"token_hash"`
+	RelatesTo string `json:"relates_to" db:"relates_to"`
+
+	CreatedAt time.Time `json:"created_at" db:"created_at"`
+	UpdatedAt time.Time `json:"updated_at" db:"updated_at"`
+}
+
+func (OneTimeToken) TableName() string {
+	return "one_time_tokens"
+}
+
+func ClearAllOneTimeTokensForUser(tx *storage.Connection, userID uuid.UUID) error {
+	return tx.Q().Where("user_id = ?", userID).Delete(OneTimeToken{})
+}
+
+func ClearOneTimeTokenForUser(tx *storage.Connection, userID uuid.UUID, tokenType OneTimeTokenType) error {
+	if err := tx.Q().Where("token_type = ? and user_id = ?", tokenType, userID).Delete(OneTimeToken{}); err != nil {
```

Comment:
Nice, index on user_id and token_type covers both queries since `token_type = ? and user_id = ?` is equivalent to `user_id = ? and token_type = ?` for the postgres query optimizer

Reply:
This is really to make sure there's no conflict on insert.